patroni/dcs/__init__.py

```diff
@@ -1131,50 +1146,67 @@ def _get_members_slots(self, name: str, role: str) -> Dict[str, Dict[str, str]]:
         # also exlude members with disabled WAL streaming
         members = filter(lambda m: m.name != name and not m.nostream, self.members)
 
-        if role in ('master', 'primary', 'standby_leader'):
+        if can_advance_slots and global_config.member_slots_ttl > 0:
+            # if the node does only cascading and can't become the leader, we
+            # want only to have slots for members that could connect to it.
+            members = [m for m in members if not nofailover or m.replicatefrom == name]
+        elif role in ('master', 'primary', 'standby_leader'):  # PostgreSQL is older than 11
+            # on the leader want to have slots only for the nodes that are supposed to be replicating from it.
             members = [m for m in members if m.replicatefrom is None
                        or m.replicatefrom == name or not self.has_member(m.replicatefrom)]
         else:
             # only manage slots for replicas that replicate from this one, except for the leader among them
             members = [m for m in members if m.replicatefrom == name and m.name != self.leader_name]
 
-        slots = {slot_name_from_member_name(m.name): {'type': 'physical'} for m in members}
-        if len(slots) < len(members):
+        slots: Dict[str, int] = self.slots
+        ret: Dict[str, Dict[str, Any]] = {}
+        for member in members:
+            slot_name = slot_name_from_member_name(member.name)
+            lsn = slots.get(slot_name, 0)
+            if member.replicatefrom:
+                lsn = max(member.lsn or 0, lsn)
```

Comment:
It is again about cascading setups:
```
primary<-node1<-node2(replicatefrom: node1)
```

`/status: {"slots":{...}}` in DCS is updated only by the primary.
But, replication slot for `node2` on the primary is not active, therefore as a workaround we want to take `received/replayed` LSN reported by the `node2`. `max` is used just in case, for example if there is no feedback from the `node2` because it is restarting.
Later, `primary` will advance LSN of the `node2` slot and update "slots" value in DCS.

Reply:
Got it, thanks!

Reply:
```suggestion
            if member.replicatefrom:
                # `/status` key is maintained by the leader, but `member` may be connected to some other node. In that
                # case, the slot in the leader is inactive and doesn't advance, so we use the LSN reported by the member
                # `max` is only a fallback so we take the LSN from the slot when there is no feedback from the member
                lsn = max(member.lsn or 0, lsn)
```

I wonder if this would be worth it?